### PR TITLE
[libc][math][c23] Enable C23 _Float16 math functions on GPUs

### PR DESCRIPTION
**Pull request by overmighty**

### CODE REVIEW

**Comment by overmighty** [comment]
@jhuber6 I built the tests for AMDGPU, but only ran them on NVIDIA. Could you please try to run them on AMD for me, or should I try to run them on my other laptop?

**Review by jhuber6 — APPROVED** [review]
Works on both targets, thanks. I'm assuming these implementations will go to the builtins where appropriate?

**Comment by overmighty** [comment]
Oh, they actually don't. I forgot to add an `OR LIBC_TARGET_OS_IS_GPU` in https://github.com/llvm/llvm-project/blob/main/libc/cmake/modules/LLVMLibCFlagRules.cmake#L280-L281.

**Comment by overmighty** [comment]
Should we try to run some benchmarks before enabling builtins on GPUs?

**Comment by jhuber6** [comment]
> Should we try to run some benchmarks before enabling builtins on GPUs?

If the builitin is supported it's pretty much always going to be faster than the bit-twidding generic implementation. Maybe there's some exceptions for things like `sin` or `sqrt` which might have precision issues. @arsenm the f16 builtins are generally supported in hardware, right?

**Comment by arsenm** [comment]
> > Should we try to run some benchmarks before enabling builtins on GPUs?
> 
> If the builitin is supported it's pretty much always going to be faster than the bit-twidding generic implementation. Maybe there's some exceptions for things like `sin` or `sqrt` which might have precision issues. @arsenm the f16 builtins are generally supported in hardware, right?

sqrt is always correct. The intrinsics should all be lowered correctly. If not, it's a bug that should be addressed in the backend and not worked around in a library. I'm not sure if the sin/cos cases pass conformance, OpenCL only just finally added half tests to the conformance suite so we can check. The f16 cases are easier because it's a lot easier to upcast to float and back

File: libc/src/math/generic/CMakeLists.txt
Comment (lntue) [inline libc/src/math/generic/CMakeLists.txt:114]: Do you mind running the bazel builds to see if these extra dependencies need to be added to bazel overlay?

Comment (overmighty) [reply libc/src/math/generic/CMakeLists.txt:114]: Bazel build seems to work:

```
~/projects/llvm-project/utils/bazel libc-math-float16-gpu ➜ bazel build --config=generic_clang @llvm-project//libc/...
INFO: Analyzed 897 targets (68 packages loaded, 6253 targets configured).
INFO: Found 897 targets...
INFO: Elapsed time: 166.512s, Critical Path: 151.57s
INFO: 3876 processes: 2132 internal, 1744 linux-sandbox.
INFO: Build completed successfully, 3876 total actions
~/projects/llvm-project/utils/bazel libc-math-float16-gpu ➜ bazel test --config=generic_clang @llvm-project//libc/...
INFO: Analyzed 897 targets (0 packages loaded, 0 targets configured).
INFO: Found 659 targets and 238 test targets...
INFO: Elapsed time: 5.366s, Critical Path: 4.26s
INFO: 239 processes: 476 linux-sandbox.
INFO: Build completed successfully, 239 total actions
@llvm-project//libc/test/src/__support:arg_list_test                     PASSED in 0.0s
...

Executed 238 out of 238 tests: 238 tests pass.
There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see which ones these are.
~/projects/llvm-project/utils/bazel libc-math-float16-gpu ➜
```

Why are only some math functions listed in https://github.com/llvm/llvm-project/blob/main/utils/bazel/llvm-project-overlay/libc/BUILD.bazel?

Comment (lntue) [reply libc/src/math/generic/CMakeLists.txt:114]: B/c I've been lazy / forget to update it, especially since there are quite some work need to be done to add smoke tests to the lists.  I planned to update it all at once at some point in late July, early August.

**Comment by lntue** [comment]
The PR looks good to me.  Let's wait for @jhuber6 to do the confirmation / approval.

**Review by jhuber6 — APPROVED** [review]

**Review by lntue — APPROVED** [review]

**Comment by overmighty** [comment]
https://github.com/llvm/llvm-project/pull/98972 was merged yesterday so maybe I should enable these functions in this PR too while at it, unless you would prefer another PR as this one is already approved.

**Comment by overmighty** [comment]
Rebased to enable `f16mul` and `f16mulf`.

**Comment by overmighty** [comment]
@jhuber6 Is this still okay? I just disabled the functions that use `long double` and enabled `f16mul` and `f16mulf`.

**Comment by jhuber6** [comment]
> @jhuber6 Is this still okay? I just disabled the functions that use `long double` and enabled `f16mul` and `f16mulf`.

Yes, so long as there's no long double stuff it should still work. If not we can turn it off, but it worked when I last tested it.

**Review by jhuber6 — APPROVED** [review]

**Pull request merged**
